libs/test-utils/src/mock_function.ts

```diff
@@ -114,22 +128,69 @@ export function mockFunction<Func extends AnyFunc>(
     actualCalls: [],
   };
 
+  // Find the next expected call that should be used for the given actual call.
+  // We track which expected call each actual call matched and then base the
+  // next appropriate expected call on that.
+  function findExpectedCallForActualCallIndex(actualCallIndex: number) {
+    const previousActualCall = state.actualCalls[actualCallIndex - 1];
+    if (!previousActualCall) {
+      return state.expectedCalls[0];
+    }
+
+    // If the previous actual call didn't match any expected call, then we don't
+    // know what the next expected call should be.
+    if (previousActualCall.matchingExpectedCallIndex === undefined) {
+      throw new MockFunctionError(
+        `Previous call to mock function did not match expected calls`
+      );
+    }
+
+    // If we have a next expected call, use that
+    const nextExpectedCallIndex =
+      previousActualCall.matchingExpectedCallIndex + 1;
+    if (nextExpectedCallIndex < state.expectedCalls.length) {
+      return state.expectedCalls[nextExpectedCallIndex];
+    }
+
+    // Otherwise, maybe the previous expected call was a repeated expected call,
+    // so we can use that
+    const previousExpectedCall =
+      state.expectedCalls[previousActualCall.matchingExpectedCallIndex];
+    if (assertDefined(previousExpectedCall).repeated) {
+      return previousExpectedCall;
+    }
```

Comment:
I'm having a little trouble gleaning what's supported looking at this logic, but is a repeated call allowed to be interleaved with non-repeated calls, e.g.

Given the following expectations:
- Expect repeated calls for auth status
- Expect call to endpoint 1
- Expect call to endpoint 2

Is this sequence valid?
- Call for auth status
- Call to endpoint 1
- Call for auth status
- Call to endpoint 2

We're probably in agreement that this should be valid. I'm just having a hard time seeing whether this is supported looking at this logic

Reply:
Ah yes of course - thanks for clarifying that! Lost sight of that as I was in the weeds thinking about the current code block